### ISSUE TITLE
[RAFT] Directly decrease pr.next when receive rejection from follower

### ISSUE THREAD

**Issue by shenlanse** [etcd-io/etcd#8393]
### My Confusion
when leader receives rejection from follower, leader should  decrease corresponding progress.next.    

I think when in ProgressStateReplicate, we should set pr.Next = pr.last + 1.  I don't know the reason why we set pr.Next = pr.Match + 1 instead.

Message.RejectHint indicates largest index the follower has. So RejectHint+1 is exactly what follower expects.

[https://github.com/coreos/etcd/blob/master/raft/progress.go#L139](https://github.com/coreos/etcd/blob/master/raft/progress.go#L139)

```
func (pr *Progress) maybeDecrTo(rejected, last uint64) bool {
	if pr.State == ProgressStateReplicate {
		// the rejection must be stale if the progress has matched and "rejected"
		// is smaller than "match".
		if rejected <= pr.Match {
			return false
		}
		// directly decrease next to match + 1
		pr.Next = pr.Match + 1
		return true
	}

	// the rejection must be stale if "rejected" does not match next - 1
	if pr.Next-1 != rejected {
		return false
	}

	if pr.Next = min(rejected, last+1); pr.Next < 1 {
		pr.Next = 1
	}
	pr.resume()
	return true
}
```

**Comment by xiang90** [comment]
> pr.Next = pr.last + 1

what is pr.last?

**Comment by shenlanse** [comment]
Sorry for mistake. 
`pr.Next=last  + 1`
that is to say:
`pr.Next= RejectHint + 1`

**Comment by xiang90** [comment]
@shenlanse We probably could do that. I do not really remember the reason for decreasing it directly to match. probably it is because rejected is not even introduced at that time. you might want to figure it out by looking through git history. i will give it more thoughts too.

**Comment by shenlanse** [comment]
@xiang90  all  right, thanks

**Comment by shenlanse** [comment]
I think maybe because:
RejectHint is the last index of follower, but the corresponding term of the last entry might is not same with current leader. 

If we set pr.Next=RejectHint + 1, then when leader send messages again to this follower according to progress, matchTerm might return false, so follower will reject again with same last index.

But if we set pr.Next=pr.Match +1, follower will accept it (matchTerm return true) definitely.

If what i said above is right, then we can optimize the algo. when follower rejects the entry, it will not only return RejectHint but also return term of last entry. So when leader get this response, it can make a decision:
(1) set pr.Next=last + 1
(2) set pr.Next=pr.Match + 1
